packages/utils/src/types/publicKey.ts

```diff
@@ -0,0 +1,142 @@
+import { publicKeyConvert, publicKeyCreate, ecdsaRecover } from 'secp256k1'
+import { u8aToHex, u8aEquals, stringToU8a } from '../u8a'
+import PeerId from 'peer-id'
+import { pubKeyToPeerId } from '../libp2p'
+import { Address, Hash } from './primitives'
+
+export class PublicKey {
+  // Cache expensive computation result
+  private _address: Address
+
+  private constructor(private arr: Uint8Array) {}
+
+  static fromPrivKeyString(privKey: string) {
+    return PublicKey.fromPrivKey(stringToU8a(privKey))
+  }
+
+  static fromPrivKey(privKey: Uint8Array): PublicKey {
+    if (privKey.length !== 32) {
+      throw new Error('Incorrect size Uint8Array for private key')
+    }
+
+    return new PublicKey(publicKeyCreate(privKey, false))
+  }
+
+  static deserialize(arr: Uint8Array) {
+    switch (arr.length) {
+      case 65:
+        if (arr[0] != 4) {
+          throw Error(`Invalid uncompressed public key`)
+        }
+        return new PublicKey(arr)
+      case 64:
+        return new PublicKey(Uint8Array.from([4, ...arr]))
+      case 33:
+        if (![2, 3].includes(arr[0])) {
+          throw Error(`Invalid compressed public key`)
+        }
+        return new PublicKey(arr)
+      default:
+        throw Error(`Invalid length ${arr.length} of public key`)
+    }
+  }
+
+  static fromPeerId(peerId: PeerId): PublicKey {
+    return PublicKey.deserialize(peerId.pubKey.marshal())
+  }
+
+  static fromPeerIdString(peerIdString: string) {
+    return PublicKey.fromPeerId(PeerId.createFromB58String(peerIdString))
+  }
+
+  static fromSignature(hash: string, r: string, s: string, v: number): PublicKey {
+    return new PublicKey(
+      ecdsaRecover(Uint8Array.from([...stringToU8a(r), ...stringToU8a(s)]), v, stringToU8a(hash), false)
+    )
+  }
+
+  static fromString(str: string): PublicKey {
+    if (!str || str.length == 0) {
+      throw new Error('Cannot determine address from empty string')
+    }
+    return PublicKey.deserialize(stringToU8a(str))
+  }
+
+  static get SIZE_COMPRESSED(): number {
+    return 33
+  }
+
+  static get SIZE_UNCOMPRESSED(): number {
+    return 65
+  }
+
+  public get isCompressed(): boolean {
+    return [2, 3].includes(this.arr[0])
```

Comment:
Uncompressed public keys can have first byte set to `0x04` as well - see https://github.com/bitcoin-core/secp256k1/blob/e0508ee9db2725c5efbc1cad434e14c2ebf55427/src/eckey_impl.h#L49 and https://github.com/bitcoin-core/secp256k1/blob/e0508ee9db2725c5efbc1cad434e14c2ebf55427/src/eckey_impl.h#L21

Reply:
Absolutely, found it more intuitive to have a property `isCompressed` that works without sth `return !isCompressed` - shouldn't matter whether you test whether the first bit is `0x04` or `0x02`/`0x03`